### PR TITLE
Lot of fixies, and improvements (which should not be in one PR, but now it's too late)

### PR DESCRIPTION
**Pull request by outscale-mgo**

### CODE REVIEW

File: tests/06_multiple_ipv4_on_nic/test.sh [hunk 1]
```diff
@@ -46,6 +46,8 @@ ssh_ping_ip 2 $b3 $a1
 ssh_ping_ip 2 $b3 $a2
 ssh_ping_ip 2 $b3 $a3
 
+
```
Comment (outscale-toa) [inline tests/06_multiple_ipv4_on_nic/test.sh:50]: can you remove the white space please ?

File: api/client/client.cc [hunk 2]
```diff
@@ -34,23 +36,30 @@ GlobalOptions::GlobalOptions() {
...
             verbose = true;
-        else if (string(argv[i]) == "-V" ||
-                 string(argv[i]) == "--version")
+        } else if (string(argv[i]) == "-V" ||
+                 string(argv[i]) == "--version") {
             version = true;
-        else if (string(argv[i]) == "-h" ||
-                 string(argv[i]) == "--help")
+        } else if (string(argv[i]) == "-h" ||
+                   string(argv[i]) == "--help") {
```
Comment (outscale-fne) [inline api/client/client.cc:55]: White space at the beginning of the line... Wrong indent :)

File: api/client/nic.cc [hunk 3]
```diff
@@ -407,20 +413,34 @@ static inline bool CheckOption(int count, int argc, char **argv,
     return count + 1 < argc && string(argv[count]) == option;
 }
 
+namespace {
+
+string mk_opt_option(const char *f, char *opt) {
+    return string(f) + ": \""+ opt + "\"";
+}
```
Comment (outscale-fne) [inline api/client/nic.cc:420]: Why not incrementing indentation here?

File: api/common/vec.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        V_FOREACH_(vec, val, condition, __LINE__)
+
+#define V_CAT_(a,b) a ## b
+
+#define V_FOREACH_IF_(vec, val, if_cnd, u_cnd, l)                       \
+        type(vec) V_CAT_(val, l) = *&vec;                             \
+        for (type((*vec) + 0) val = *vec;                             \
+             ({  do {val = *V_CAT_(val, l)++;}                          \
+                     while (!(if_cnd));					\
+                     val;                                               \
```
Comment (outscale-fne) [inline api/common/vec.h]: Could be prettier if all `\` are in the same column...

File: api/protocol/message_0.proto [hunk 5]
```diff
@@ -188,6 +189,13 @@ message MessageV0 {
     optional bool packet_trace = 11;
     //path to store pcap file
     optional string packet_trace_path = 12;
+    //if type is benchmark, the kind of bebchmark you want to do
```
Comment (outscale-fne) [inline api/protocol/message_0.proto]: *benchmark

File: api/server/app.h [hunk 6]
```diff
@@ -125,6 +125,7 @@ extern Model model;
 extern Log log;
 extern Graph graph;
 extern pg_error *pg_error;
+
```
Comment (outscale-fne) [inline api/server/app.h:128]: Single new blank line?

File: api/server/graph.cc [hunk 7]
```diff
@@ -56,6 +62,210 @@ void BuildMulticastIp6(uint8_t *multicast_ip, uint32_t vni) {
...
+
+    gettimeofday(&tv, NULL);
+    t = 1000000 * tv.tv_sec + tv.tv_usec;
+    return t;
+}
+
+void rx_snd_callback(struct pg_brick *b, pg_packet_t **burst,
+                     uint16_t len, void *private_data) {
+    Graph::BenchSnd *bsnd = reinterpret_cast<Graph::BenchSnd *>(private_data);
+    // printf("rx_snd_callback - %d - %d\n", bsnd->not_droped, len);
```
Comment (outscale-fne) [inline api/server/graph.cc]: Commented code?

File: api/server/graph.cc [hunk 8]
```diff
@@ -56,6 +62,210 @@ void BuildMulticastIp6(uint8_t *multicast_ip, uint32_t vni) {
...
+    if (!len)
+        return;
+
+    // printf("pong\n");
```
Comment (outscale-fne) [inline api/server/graph.cc]: Debug-related printf?

File: api/server/graph.cc [hunk 9]
```diff
@@ -56,6 +62,210 @@ void BuildMulticastIp6(uint8_t *multicast_ip, uint32_t vni) {
...
+
+
+void tx_snd_callback(struct pg_brick *b, pg_packet_t **burst,
+                     uint16_t *len, void *private_data) {
+    Graph::BenchSnd *bsnd = reinterpret_cast<Graph::BenchSnd *>(private_data);
+    Graph::BenchInfo *binfo = &bsnd->binfo;
+    static uint16_t seq;
+    uint32_t time = get_time();
+
+    // printf("Go %x !!!\n", binfo->ipdest);
```
Comment (outscale-fne) [inline api/server/graph.cc]: Debug-related printf?

File: api/server/graph.cc [hunk 10]
```diff
@@ -304,12 +515,14 @@ void *Graph::Poller(void *graph) {
...
+        if (unlikely(pg_brick_poll(nic, &pkts_count, &poll_err) < 0)) {
+            PG_ERROR_(poll_err);
+        }
+
         for (uint32_t v = 0; v < size; v++) {
-            if (pg_brick_poll(list->pollables[v],
-                              &pkts_count, &app::pg_error) < 0) {
-                PG_ERROR_(app::pg_error);
+            if (unlikely(pg_brick_poll(list->pollables[v],
+                                       &pkts_count, &poll_err) < 0)) {
```
Comment (outscale-fne) [inline api/server/graph.cc:516]: Indentation error...

**Review by outscale-fne — CHANGES_REQUESTED** [review]
J'ai un peu titillé! Mais c'est surtout du formattage...

File: api/client/client.cc (hunk 2 above)
Comment (outscale-mgo) [reply api/client/client.cc:55]: I think in this case, the format was wrong before this code.
also, the white-space are not at the beginning of the line but after and before the newly added braces

File: api/client/nic.cc (hunk 3 above)
Comment (outscale-mgo) [reply api/client/nic.cc:420]: because google style ask not to

**Review by outscale-toa — APPROVED** [review]

File: tests/functions.sh [hunk 11]
```diff
@@ -682,10 +682,33 @@ function server_start_bonding {
...
+    fi
+    if [ $? -ne 0 ]; then
+        sudo kill -2 $(ps --ppid ${server_pids[$id]} -o pid=) &> /dev/null
+        sleep 0.1
+        sudo kill -s 0 ${server_pids[$id]} &> /dev/null
+    fi
+    if [ $? -ne 0 ]; then
+        sudo kill -15 $(ps --ppid ${server_pids[$id]} -o pid=) &> /dev/null
+        sleep 0.4
+        sudo kill -s 0 ${server_pids[$id]} &> /dev/null
```
Comment (outscale-toa) [inline tests/functions.sh]: Can you do a loop or function to kill server_pids please?

File: tests/functions.sh [hunk 12]
```diff
@@ -647,7 +647,7 @@ function server_start_options {
     options=${@:2}
     echo "[butterfly-$id] starting"
 
-    exec sudo $BUTTERFLY_BUILD_ROOT/api/server/butterflyd --dpdk-args "--no-shconf -c1 -n1 --vdev=eth_pcap$id,iface=but$id --no-huge" -l debug -i ::101 -s /tmp --endpoint=tcp://0.0.0.0:876$id $options &> $BUTTERFLY_BUILD_ROOT/butterflyd_${id}_output &
+    exec sudo $BUTTERFLY_BUILD_ROOT/api/server/butterflyd --dpdk-args "--no-shconf -c1 -n1 --vdev=eth_pcap$id,iface=but$id --no-huge" -l debug -i ::101 -s /tmp --endpoint=tcp://0.0.0.0:876$id $options &#> $BUTTERFLY_BUILD_ROOT/butterflyd_${id}_output &
```
Comment (outscale-toa) [inline tests/functions.sh]: Can you uncomment the output  file please?

File: tests/functions.sh [hunk 13]
```diff
@@ -19,6 +19,11 @@ RETURN_CODE=0
 
 function return_result {
     clean_all
+    if [ $RETURN_CODE == 0 ]; then
+	echo $CUR_TEST 'SUCESS !!!'
+    else
+	echo 'FAIL !!!'
```
Comment (outscale-toa) [inline tests/functions.sh]: Why not 	```echo $CUR_TEST 'FAIL !!!' ```

File: tests/functions.sh [hunk 14]
```diff
@@ -567,20 +600,31 @@ function qemu_start {
         CMD="sudo qemu-system-x86_64 -netdev user,id=network0,hostfwd=tcp::500${id}-:22 -device e1000,netdev=network0 -m 124M -enable-kvm  -netdev type=tap,id=mynet1,ifname=tap-q-$id,script=no,downscript=no -device virtio-net-pci,gso=off,mac=$MAC,netdev=mynet1 -drive file=$IMG_PATH -snapshot -display none"
     else
         CMD="sudo qemu-system-x86_64 -netdev user,id=network0,hostfwd=tcp::500${id}-:22 -device e1000,netdev=network0 -m 124M -enable-kvm -chardev socket,id=char0,path=$SOCKET_PATH -netdev type=vhost-user,id=mynet1,chardev=char0,vhostforce -device virtio-net-pci,csum=off,gso=off,mac=$MAC,netdev=mynet1 -object memory-backend-file,id=mem,size=124M,mem-path=/mnt/huge,share=on -numa node,memdev=mem -mem-prealloc -drive file=$IMG_PATH -snapshot -display none"
+        nic_type="VM"
     fi
-    exec $CMD &> $BUTTERFLY_BUILD_ROOT/qemu_${id}_output &
+    exec $CMD &#> $BUTTERFLY_BUILD_ROOT/qemu_${id}_output &
```
Comment (outscale-toa) [inline tests/functions.sh]: Can you uncomment the log file please?

**Review by outscale-toa — CHANGES_REQUESTED** [review]

**Pull request merged**